java/ql/src/semmle/code/java/Annotation.qll

```diff
@@ -99,18 +166,69 @@ private predicate sourceAnnotValue(Annotation a, Method m, Expr val) {
 
 /** An abstract representation of language elements that can be annotated. */
 class Annotatable extends Element {
-  /** Holds if this element has an annotation. */
-  predicate hasAnnotation() { exists(Annotation a | a.getAnnotatedElement() = this) }
+  /** Holds if this element has an annotation, including inherited annotations. */
+  predicate hasAnnotation() { exists(getAnAnnotation()) }
+
+  /** Holds if this element has a declared annotation, excluding inherited annotations. */
+  predicate hasDeclaredAnnotation() { exists(getADeclaredAnnotation()) }
 
-  /** Holds if this element has the specified annotation. */
+  /**
+   * Holds if this element has the specified annotation, including inherited
+   * annotations.
+   */
   predicate hasAnnotation(string package, string name) {
     exists(AnnotationType at | at = getAnAnnotation().getType() |
       at.nestedName() = name and at.getPackage().getName() = package
     )
   }
 
-  /** Gets an annotation that applies to this element. */
-  Annotation getAnAnnotation() { result.getAnnotatedElement() = this }
+  /**
+   * Gets an annotation that applies to this element, including inherited annotations.
+   * The results only include _direct_ annotations; _indirect_ annotations, that is
+   * repeated annotations in an (implicit) container annotation, are not included.
+   */
+  Annotation getAnAnnotation() { result = getADeclaredAnnotation() }
+
+  /**
+   * Gets an annotation that is declared on this element, excluding inherited annotations.
+   */
+  Annotation getADeclaredAnnotation() { result.getAnnotatedElement() = this }
+
+  /** Gets an _indirect_ (= repeated) annotation. */
+  // 'indirect' as defined by https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/AnnotatedElement.html
+  private Annotation getAnIndirectAnnotation() {
+    exists(AnnotationType t, Annotation containerAnn |
+      t = result.getType() and
+      containerAnn = getADeclaredAnnotation() and
+      containerAnn.getType() = t.getContainingAnnotationType()
+    |
+      result = containerAnn.getAValue("value")
+    )
+  }
+
+  private Annotation getAnAssociatedAnnotation(AnnotationType t) {
+    result.getType() = t and
+    // Direct or indirect annotation
+    if getADeclaredAnnotation().getType() = t or getAnIndirectAnnotation().getType() = t
+    then (
+      result = getADeclaredAnnotation() or result = getAnIndirectAnnotation()
+    ) else (
+      // Only if neither a direct nor an indirect annotation is present look for an inherited one
+      t.isInherited() and
+      // @Inherited only works for classes
+      result = this.(Class).getASupertype().(Class).(Annotatable).getAnAssociatedAnnotation(t)
```

Comment:
`Class` is always `Annotatable` (via `RefType`)

Reply:
You mean `(Class).(Annotatable)`, right?
The VSCode extension refuses directly using `(Class).getAnAssociatedAnnotation(t)`, I assume because the predicate is private?

Reply:
Ah yes, I suppose so, never mind then